sdk/identity/azure-identity/azure/identity/_internal/aad_client_base.py

```diff
@@ -83,24 +83,57 @@ def obtain_token_by_refresh_token(self, scopes, refresh_token, **kwargs):
     def _build_pipeline(self, config=None, policies=None, transport=None, **kwargs):
         pass
 
-    def _process_response(self, response, scopes, now):
-        # type: (dict, Sequence[str], int) -> AccessToken
-        _raise_for_error(response)
-
-        # TokenCache.add mutates the response. In particular, it removes tokens.
-        response_copy = copy.deepcopy(response)
-
-        self._cache.add(event={"response": response, "scope": scopes, "client_id": self._client_id}, now=now)
-        if "expires_on" in response_copy:
-            expires_on = int(response_copy["expires_on"])
-        elif "expires_in" in response_copy:
-            expires_on = now + int(response_copy["expires_in"])
+    def _process_response(self, response, request_time):
+        # type: (PipelineResponse, int) -> AccessToken
+
+        content = ContentDecodePolicy.deserialize_from_http_generics(response.http_response)
+
+        if response.http_request.body.get("grant_type") == "refresh_token":
+            if content.get("error") == "invalid_grant":
+                # the request's refresh token is invalid -> evict it from the cache
+                cache_entries = self._cache.find(
+                    TokenCache.CredentialType.REFRESH_TOKEN,
+                    query={"secret": response.http_request.body["refresh_token"]},
+                )
+                for invalid_token in cache_entries:
+                    self._cache.remove_rt(invalid_token)
```

Comment:
I think it will be great if we can encapsulate more on cache operations. e.g. combine find + remove_rt  & find + update_rt. But we can do it in future releases.